### ISSUE TITLE
Hibernate 5.x with Payara and Oracle driver throws reflection exception

### ISSUE THREAD

**Issue by ubench-tdv** [payara/Payara#730]
When using hibernate 5.x with Payara and a connectionpool to an Oracle database (ojdbc7.jar) you"ll see an exception:

```java
javax.resource.ResourceException: java.lang.IllegalAccessException: Class 
com.sun.gjc.util.MethodExecutor can not access a member of class 
    oracle.jdbc.driver.PhysicalConnection with modifiers "public"
    	at com.sun.gjc.util.MethodExecutor.invokeMethod(MethodExecutor.java:240)
	at com.sun.gjc.spi.jdbc40.ConnectionHolder40.getSchema(ConnectionHolder40.java:609)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.hibernate.engine.jdbc.env.internal.DefaultSchemaNameResolver$SchemaNameResolverJava17Delegate.resolveSchemaName(DefaultSchemaNameResolver.java:72)
	at org.hibernate.engine.jdbc.env.internal.DefaultSchemaNameResolver.resolveSchemaName(DefaultSchemaNameResolver.java:59)
	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentImpl.determineCurrentSchemaName(JdbcEnvironmentImpl.java:294)`
```

Digging a little deeper this is a bug in com.sun.gjc.spi.jdbc40.ConnectionHolder40

```java
public String getSchema() throws SQLException {
    if(DataSourceObjectBuilder.isJDBC41()) {
            checkValidity();
            try {
                return (String) getMethodExecutor().invokeMethod(con, "getSchema", null);
            } catch (ResourceException ex) {
                _logger.log(Level.SEVERE, "jdbc.ex_connection_holder", ex);
                throw new SQLException(ex);
            }
        }
        throw new UnsupportedOperationException("Operation not supported in this runtime.");
}
```

Which passes the connection (internal Oracle T4CConnection class) straight to com.sun.gjc.util.MethodExcutor:

```java
public Object invokeMethod(Object object, String methodName, 
            Class<?>[] valueTypes, Object... values) throws ResourceException {
        Object returnValue = null;
        Method actualMethod = null;
        try {
            actualMethod = object.getClass().getMethod(methodName, valueTypes);
        } catch (NoSuchMethodException ex) {
            throw new ResourceException(ex);
        } catch (SecurityException ex) {
            throw new ResourceException(ex);
        }
        if (actualMethod != null) {
            try {
                returnValue = actualMethod.invoke(object, values);
            } catch (IllegalAccessException ex) {
                throw new ResourceException(ex);
            } catch (IllegalArgumentException ex) {
                throw new ResourceException(ex);
            } catch (InvocationTargetException ex) {
                throw new ResourceException(ex);
            }
        }
        return returnValue;
    }
```

Now the problem is that the actualMethod is initialized as follows:

```java
actualmethod = public java.lang.String oracle.jdbc.driver.PhysicalConnection.getSchema() throws java.sql.SQLException'
```

This is a problem since on oracle.jdbc.driver.PhysicalConnection the getSchema method is on a protected class and so we can not access it using reflection,

This is not Oracle's fault, neither Hibernate's fault since the resolving should be done on java.sql.Connection instead of oracle.jdbc.driver.PhysicalConnection.

As a bug fix, the invokeMethod of com.sun.gjc.util.MethodExcutor can have an extra parameter e.g. myInterfaceVar.

So that 

```java
actualMethod = myInterfaceVar.getClass().getMethod(methodName, valueTypes);
```

instead of

```java
actualMethod = object.getClass().getMethod(methodName, valueTypes);
```

The getSchema method should then pass:

```java
return (String) getMethodExecutor().invokeMethod(con, java.sql.Connection, "getSchema", null);
```

**Comment by mulderbaba** [comment]
[Attached](https://github.com/payara/Payara/files/198476/hibernate5-test.zip) is the maven based reproducer project of this error. In order to use it, first you need to install ojdbc7.jar to maven local repository, which is %user_home%/.m2 folder. ojdbc7.jar can be downloaded from: http://www.oracle.com/technetwork/database/features/jdbc/jdbc-drivers-12c-download-1958347.html. (You need an OTN account to download the artifact.) Run maven command to install the jar file: mvn install:install-file -DgroupId=com.oracle -DartifactId=ojdbc7 -Dversion=12.1.0.1 -Dpackaging=jar -Dfile=/path/to/ojdbc7.jar -DgeneratePom=true

Steps to reproduce:
- In order to use hibernate5 with a connection pool, first a new JDBC Connection Pool, with vendor: Oracle, should be defined if there is none.
- After having JDBC Connection Pool with pool name: OraclePool, define a JDBC Resource with JNDI name: jdbc/hibernate5. Select OraclePool as your pool.
- Deploy hibernate5-test-0.0.1-SNAPSHOT.war generated from the attached maven project.
- You will see the error in server.log while doing the deploy.

**Comment by ubench-tdv** [comment]
Many thanks for attaching this project!

**Comment by mikecroft** [comment]
Internal defect ID - `PAYARA-717`

**Comment by smillidge** [comment]
Fixed by #762
